File: python/338_countBits.py
# n&n(n - 1) will remove the lowest "1" bit
class Solution(object):
    def countBits(self, num):
        """
        :type num: int
        :rtype: List[int]
        """
        res = [0]
        for i in xrange(1, num + 1):
            res.append(res[i&(i-1)] + 1)
        
        return res

File: C/338_countBits.c
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* countBits(int num, int* returnSize) {
    int *res = (int *)malloc(sizeof(int) * (num + 1));
    int i;
    
    *returnSize = num + 1;
    res[0] = 0;
    
    for (i = 1; i <= num; i++) {
        res[i] = res[ i & (i - 1)] + 1;
    }
    
    return res;
}
